### PR TITLE
Add Canary platform

### PR DESCRIPTION
**Pull request by jamierocks**

### CODE REVIEW

File: src/main/java/com/demonwav/mcdev/platform/canary/NeptuneModuleType.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.demonwav.mcdev.platform.canary;
+
+import org.jetbrains.annotations.NotNull;
+
+public class NeptuneModuleType extends CanaryModuleType {
+
+    private static final String ID = "PAPER_MODULE_TYPE";
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/NeptuneModuleType.java]: Nope.

Comment (jamierocks) [inline src/main/java/com/demonwav/mcdev/platform/canary/NeptuneModuleType.java]: Opps

File: src/main/java/com/demonwav/mcdev/platform/PlatformType.java [hunk 2]
```diff
@@ -21,6 +22,7 @@
     PAPER(PaperModuleType.getInstance(), "PAPER_MODULE_TYPE", "Paper"),
     FORGE(ForgeModuleType.getInstance(), "FORGE_MODULE_TYPE", "Forge"),
     SPONGE(SpongeModuleType.getInstance(), "SPONGE_MODULE_TYPE", "Sponge"),
+    CANARY(CanaryModuleType.getInstance(), "CANARY_MODULE_TYPE", "Canary"),
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/PlatformType.java]: Missing NEPTUNE_MODULE_TYPE, once it's added.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModule.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package com.demonwav.mcdev.platform.canary;
+
+import com.demonwav.mcdev.buildsystem.BuildSystem;
+import com.demonwav.mcdev.platform.AbstractModule;
+import com.demonwav.mcdev.platform.PlatformType;
+import com.demonwav.mcdev.platform.canary.util.CanaryConstants;
+import com.intellij.openapi.module.Module;
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModule.java]: Check the style guide, `com.demonwav` imports should be separate. Use [this](https://github.com/DemonWav/MinecraftDev/blob/master/intellij-java-style.xml) to set the style properly in IntelliJ easilly.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package com.demonwav.mcdev.platform.canary;
+
+import com.demonwav.mcdev.platform.AbstractModuleType;
+import com.demonwav.mcdev.platform.PlatformType;
+import com.demonwav.mcdev.platform.canary.util.CanaryConstants;
+import com.demonwav.mcdev.util.CommonColors;
+import com.google.common.collect.ImmutableList;
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: Check the style guide, `com.demonwav` imports should be separate. Use [this](https://github.com/DemonWav/MinecraftDev/blob/master/intellij-java-style.xml) to set the style properly in IntelliJ easilly.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    public String getId() {
+        return ID;
+    }
+
+    @NotNull
+    @Override
+    public List<String> getIgnoredAnnotations() {
+        return ImmutableList.of(
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: You can cache this `List` as a field instead.

File: src/main/java/com/demonwav/mcdev/platform/canary/util/CanaryLegacyColors.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package com.demonwav.mcdev.platform.canary.util;
+
+import com.demonwav.mcdev.util.CommonColors;
+import org.jetbrains.annotations.NotNull;
+
+import java.awt.Color;
+import java.util.Map;
+
+public final class CanaryLegacyColors {
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/util/CanaryLegacyColors.java]: Why have this as a class just to apply colours?

Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/util/CanaryLegacyColors.java]: Yeah, should just do the apply in the ModuleType. Doesn't need to look exactly like the other ModuleTypes.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+import javax.swing.Icon;
+
+public class CanaryModuleType extends AbstractModuleType<CanaryModule> {
+
+    private static final String ID = "CANARY_MODULE_TYPE";
+    private static final CanaryModuleType instance = new CanaryModuleType();
+
+    private CanaryModuleType() {
+        super("net.canarymod", "CanaryLib");
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: You can call `this("net.canarymod", "CanaryLib")` instead, and avoid duplicate code in the constructors.

File: src/main/java/com/demonwav/mcdev/buildsystem/maven/MavenBuildSystem.java [hunk 8]
```diff
@@ -10,9 +10,9 @@
 import com.demonwav.mcdev.platform.ProjectConfiguration;
 import com.demonwav.mcdev.platform.bukkit.BukkitTemplate;
 import com.demonwav.mcdev.platform.bungeecord.BungeeCordTemplate;
+import com.demonwav.mcdev.platform.canary.CanaryTemplate;
 import com.demonwav.mcdev.platform.sponge.SpongeTemplate;
 import com.demonwav.mcdev.util.Util;
-
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/buildsystem/maven/MavenBuildSystem.java]: Tsk tsk tsk.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    );
+    private static final List<String> LISTENER_ANNOTATIONS = ImmutableList.of(CanaryConstants.HANDLER_ANNOTATION);
+
+    private CanaryModuleType() {
+        this("net.canarymod", "CanaryLib");
+    }
+
+    protected CanaryModuleType(final String groupId, final String artifactId) {
+        super(groupId, artifactId);
+        CommonColors.applyStandardColors(this.colorMap, CanaryConstants.CHAT_FORMAT_CLASS);
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: `ChatFormat` does not use standard naming.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public PlatformType getPlatformType() {
+        return PlatformType.CANARY;
+    }
+
+    @Override
+    public Icon getIcon() {
+        return null;
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: no

File: src/main/java/com/demonwav/mcdev/util/MinecraftFileTemplateGroupFactory.java [hunk 11]
```diff
@@ -24,6 +24,9 @@
     @NotNull public static final String SPONGE_MAIN_CLASS_TEMPLATE = "sponge_main_class.java";
     @NotNull public static final String SPONGE_POM_TEMPLATE = "sponge_pom_template.xml";
 
+    @NotNull public static final String CANARY_MAIN_CLASS_TEMPLATE = "canary_main_class.java";
+    @NotNull public static final String CANARY_POM_TEMPLATE = "canary_pom_template.xml";
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/util/MinecraftFileTemplateGroupFactory.java]: Should go lower, under `LITELOADER_MAIN_CLASS_TEMPLATE `.

File: src/main/resources/META-INF/plugin.xml [hunk 12]
```diff
@@ -49,6 +49,8 @@
         <externalProjectDataService implementation="com.demonwav.mcdev.platform.bukkit.gradle.SpigotDataService"/>
         <externalProjectDataService implementation="com.demonwav.mcdev.platform.bukkit.gradle.PaperDataService"/>
         <externalProjectDataService implementation="com.demonwav.mcdev.platform.bungeecord.gradle.BungeeCordDataService"/>
+        <externalProjectDataService implementation="com.demonwav.mcdev.platform.canary.gradle.CanaryDataService"/>
+        <externalProjectDataService implementation="com.demonwav.mcdev.platform.canary.gradle.NeptuneDataService"/>
```
Comment (kashike) [inline src/main/resources/META-INF/plugin.xml]: Move both of these to under `<externalProjectDataService implementation="com.demonwav.mcdev.platform.liteloader.gradle.LiteLoaderDataService"/>`.

File: src/main/resources/META-INF/plugin.xml [hunk 13]
```diff
@@ -205,6 +215,8 @@
         <importer implementation="com.demonwav.mcdev.platform.bukkit.maven.PaperMavenImporter"/>
         <importer implementation="com.demonwav.mcdev.platform.bukkit.maven.PaperMavenImporter"/>
         <importer implementation="com.demonwav.mcdev.platform.bungeecord.maven.BungeeCordMavenImporter"/>
+        <importer implementation="com.demonwav.mcdev.platform.canary.maven.CanaryMavenImporter"/>
+        <importer implementation="com.demonwav.mcdev.platform.canary.maven.NeptuneMavenImporter"/>
```
Comment (kashike) [inline src/main/resources/META-INF/plugin.xml]: Both to under `<importer implementation="com.demonwav.mcdev.platform.sponge.maven.SpongeMavenImporter"/>`.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java (hunk 9 above)
Comment (jamierocks) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: Does it not?

Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: Nope.

Comment (jamierocks) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: Which colours don't match up?

Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: Sorry, appears I misread something two times. o_.

Comment (jamierocks) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: That's alright ;)

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java [hunk 14]
```diff
@@ -46,4 +49,34 @@ public static String applyPomTemplate(@NotNull Project project,
...
+        }
+
+        properties.setProperty("MAIN", settings.mainClass);
+
+        if (settings.hasAuthors()) {
+            properties.setProperty("AUTHOR_LIST", settings.authors.toString());
+            properties.setProperty("HAS_AUTHOR_LIST", "true");
+        }
+
+        // TODO: Does Canary support description?
```
Comment (kashike) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: https://github.com/CanaryModTeam/CanaryLib/blob/master/src/main/java/net/canarymod/plugin/PluginDescriptor.java#L20-L30
No.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryProjectConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+                String[] files = this.mainClass.split("\\.");
+                String className = files[files.length - 1];
+
+                String packageName = this.mainClass.substring(0, this.mainClass.length() - className.length() - 1);
+                file = getMainClassDirectory(files, file);
+
+                VirtualFile mainClassFile = file.findOrCreateChildData(this, className + ".java");
+                CanaryTemplate.applyMainClassTemplate(project, mainClassFile, packageName, className);
+
+                VirtualFile pluginYml = buildSystem.getResourceDirectories().get(0).findOrCreateChildData(this, "Canary.inf");
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryProjectConfiguration.java]: pluginYml?

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    @NotNull
+    @Override
+    public List<String> getListenerAnnotations() {
+        return LISTENER_ANNOTATIONS;
+    }
+
+    @NotNull
+    @Override
+    public CanaryModule generateModule(Module module) {
+        return new CanaryModule(module, this);
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModuleType.java]: This should be `return new CanaryModule<>(module, this);`.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+            // TODO what to do when this fails?
+            e.printStackTrace();
+            return "";
+        }
+    }
+
+    public static void applyPluginDescriptionFileTemplate(@NotNull Project project,
+            @NotNull VirtualFile file,
+            @NotNull CanaryProjectConfiguration settings,
+            @NotNull BuildSystem buildSystem) {
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: These should be in line with the top `@NotNull`, like the examples in `AbstractTemplate`.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        try {
+            applyTemplate(project, file, MinecraftFileTemplateGroupFactory.CANARY_MAIN_CLASS_TEMPLATE, properties);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @NotNull
+    public static String applyPomTemplate(@NotNull Project project,
+            @NotNull String version) {
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: This should be in line with the top `@NotNull`, like the examples in `AbstractTemplate`.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.io.IOException;
+import java.util.Properties;
+
+public class CanaryTemplate extends AbstractTemplate {
+
+    public static void applyMainClassTemplate(@NotNull Project project,
+            @NotNull VirtualFile file,
+            @NotNull String packageName,
+            @NotNull String className) {
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: These should be in line with the top `@NotNull`, like the examples in `AbstractTemplate`.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+        properties.setProperty("BUILD_VERSION", version);
+
+        FileTemplateManager manager = FileTemplateManager.getInstance(project);
+        FileTemplate fileTemplate = manager.getJ2eeTemplate(MinecraftFileTemplateGroupFactory.CANARY_POM_TEMPLATE);
+        try {
+            return fileTemplate.getText(properties);
+        } catch (IOException e) {
+            // TODO what to do when this fails?
+            e.printStackTrace();
+            return "";
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: This should return null, and the method should not be marked as `@NotNull`.

File: src/main/java/com/demonwav/mcdev/platform/canary/NeptuneModuleType.java [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+import javax.swing.Icon;
+
+public class NeptuneModuleType extends CanaryModuleType {
+
+    private static final NeptuneModuleType instance = new NeptuneModuleType();
+
+    private static final String ID = "NEPTUNE_MODULE_TYPE";
+
+    protected NeptuneModuleType() {
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/NeptuneModuleType.java]: Why is this `protected`? What will extend this?

File: src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+            }
+        };
+    }
+
+    @Override
+    public BaseInspectionVisitor buildVisitor() {
+        return new BaseInspectionVisitor() {
+            @Override
+            public void visitClass(PsiClass aClass) {
+                PsiMethod[] methods = aClass.getMethods();
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java]: Should be final.

File: src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+                        isEventHandler = true;
+                        break;
+                    }
+                }
+
+                if (!isEventHandler) {
+                    return;
+                }
+
+                boolean inError = !McPsiUtil.extendsOrImplementsClass(aClass, CanaryConstants.LISTENER_CLASS);
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java]: Should be final.

File: src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java (hunk 22 above)
Comment (jamierocks) [inline src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java]: This was copy-pasted from Bukkit's.

Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/inspection/CanaryListenerImplementedInspection.java]: I need to fix Bukkit's code, then lol. But this isn't really a problem, just being super nit-picky.

File: src/main/resources/fileTemplates/j2ee/canary_pom_template.xml.ft [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <minimizeJar>true</minimizeJar>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+    <resources>
```
Comment (kashike) [inline src/main/resources/fileTemplates/j2ee/canary_pom_template.xml.ft]: This block isn't required, this is default maven behaviour.

Comment (jamierocks) [inline src/main/resources/fileTemplates/j2ee/canary_pom_template.xml.ft]: I'm pretty sure its needed for filtering. Anyhow this was copied from Bukkit's again :p

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java (hunk 20 above)
Comment (jamierocks) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: This was taken from Bukkit's one - has this changed?

Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryTemplate.java]: Ah you're right. I'll change that. All this legacy code needs to be cleaned up, and you're helping find the mistakes haha.

File: src/main/java/com/demonwav/mcdev/asset/PlatformAssets.java [hunk 25]
```diff
@@ -51,6 +51,11 @@
     @NotNull public static final Icon MIXIN_ICON_DARK = loadIcon("/assets/icons/platform/Mixins_dark.png");
     @NotNull public static final Icon MIXIN_ICON_2X_DARK = loadIcon("/assets/icons/platform/Mixins@2x_dark.png");
 
+    @NotNull public static final Icon CANARY_ICON = loadIcon("/assets/icons/platform/Canary.png");
+    @NotNull public static final Icon CANARY_ICON_2X = loadIcon("/assets/icons/platform/Canary@2x.png");
+    @NotNull public static final Icon NEPTUNE_ICON = loadIcon("assets/icons/platform/Neptune.png");
+    @NotNull public static final Icon NEPTUNE_ICON_2X = loadIcon("assets/icons/platform/Neptune@2x.png");
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/asset/PlatformAssets.java]: This causes an error, needs to start with a `/`.

**Comment by DemonWav** [comment]
Event Listener generation doesn't work.

File: src/main/java/com/demonwav/mcdev/platform/canary/CanaryModule.java [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+    public boolean isEventClassValid(PsiClass eventClass, PsiMethod method) {
+        return CanaryConstants.HOOK_CLASS.equals(eventClass.getQualifiedName());
+    }
+
+    @Override
+    public String writeErrorMessageForEventParameter(PsiClass eventClass, PsiMethod method) {
+        return "Parameter is not a subclass of " + CanaryConstants.HOOK_CLASS + "\n" +
+                "Compiling and running this listener may result in a runtime exception";
+    }
+
```
Comment (DemonWav) [inline src/main/java/com/demonwav/mcdev/platform/canary/CanaryModule.java]: Need to override `generateEventListenerMethod` for event listener generation to work.

**Comment by DemonWav** [comment]
I've tested everything I can think to test, will merge now. Won't squash, so your commits will stay.

**Pull request merged**